### ISSUE TITLE
keyboard navigation doesn't work when bundles are refreshing

### ISSUE THREAD

**Issue by percyliang** [codalab/codalab-worksheets#244]
To reproduce: create a list of ~10 bundles.  Run a bundle that takes a bit a time (e.g., 'sleep 100').  Then try to use 'k' and 'j' to move around.  When the page refreshes automatically, hitting 'k' jumps up many bundles and hitting 'j' causes things to get stuck.

**Comment by maxwang7** [comment]
I've repro'd it, it's an odd one. I'll start digging into the code, @kashizui have an idea where I should start looking for this?

**Comment by maxwang7** [comment]
I'm at worksheet_interface.jsx:setFocus(). Seems like this function has something to do with how we're setting the indices.

**Comment by maxwang7** [comment]
The relevant code that updates the worksheet with new runBundles is worksheet_interface.jsx:updateRunBundles() (line 278).

**Comment by maxwang7** [comment]
My theory of what's going on:

The sequence of events are:

1. A long running run bundle is started
2. The page is re-rendered after the ajax call returns
  - refreshWorksheet will cause the focusIndex and focusSubIndex to both be the end, moving the focus to the last element in the table
  - worksheet_items renders once, table_items renders once, then worksheet_items renders a final time
3. Because of the way Mousetrap works, the command that binds with a key last is the one that's used. In this scenario, for the 'up' key, that's the binding in worksheet_items.jsx (not the one in table_items.jsx, which is the one we desire) because of the previous order of the rendering. (this order of rendering has been confirmed through console.log statements).
4. Finally, hitting 'up' causes the incorrect mouse event to fire, causing the focusIndex to switch from 0 to -1 and breaking the indices.

Why is there a 'wall' for how much the user can move down?
- If step 4 were changed so that the 'down' key was the first key hit after refreshing, then a "wall" would appear because an improper focusIndex is set. 

Ideas of solutions:

It seems like the root of the problem is that the improper key bindings are being executed. My proposal is to move all Mousetrap calls into a singleton which is context aware (i.e. checks the current React state, then makes a decision on which callback to execute). 

These are the files that use Mousetrap:
- worksheet_items.jsx
- worksheet_item_interface.jsx
- table_item_interface.jsx
- worksheet_interface.jsx
- ws_mixins.jsx

**Comment by kashizui** [comment]
done
